pkg/auto/sccexporter/device.go

```diff
@@ -0,0 +1,122 @@
+package sccexporter
+
+import (
+	"context"
+	"encoding/json"
+
+	"go.uber.org/zap"
+
+	"github.com/smart-core-os/sc-api/go/traits"
+	"github.com/smart-core-os/sc-bos/pkg/gen"
+	"github.com/smart-core-os/sc-golang/pkg/trait"
+)
+
+// DataFetcher is a function that fetches device data for a specific trait
+type DataFetcher func(ctx context.Context) (string, error)
+
+// MeterReadingWithUnits extends gen.MeterReading with unit information.
+// This allows us to efficiently add usageUnit and producedUnit fields
+// without multiple marshal/unmarshal cycles.
+type MeterReadingWithUnits struct {
+	*gen.MeterReading
+	UsageUnit    string `json:"usageUnit,omitempty"`
+	ProducedUnit string `json:"producedUnit,omitempty"`
+}
+
+type device struct {
+	name     string
+	logger   *zap.Logger
+	traits   map[trait.Name]DataFetcher
+	info     map[string]string
+	metaData *traits.Metadata
+}
+
+func newDevice(name string, logger *zap.Logger, metaData *traits.Metadata) *device {
+	d := &device{
+		name:     name,
+		logger:   logger,
+		traits:   make(map[trait.Name]DataFetcher),
+		info:     make(map[string]string),
+		metaData: metaData,
+	}
+	return d
+}
+
+func (d *device) getMeterData(ctx context.Context, meterClient gen.MeterApiClient) (string, error) {
+	reading, err := meterClient.GetMeterReading(ctx, &gen.GetMeterReadingRequest{
+		Name: d.name,
+	})
+	if err != nil {
+		d.logger.Error("failed to get meter reading", zap.String("meter", d.name), zap.Error(err))
+		return "", err
+	}
+
+	// Use struct embedding to efficiently add unit fields in a single marshal operation
+	data := &MeterReadingWithUnits{
+		MeterReading: reading,
```

Comment:
I just noticed that this (and other d.getFoo methods) will use json instead of protojson for marshalling. That's probably not what we want.